File: src/temporal_xray/tools/get_workflow_stack_trace.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

from ..temporal_client import get_temporal_client
from ..types import GetWorkflowStackTraceInput, StackTraceOutput

PENDING_ACTIVITY_STATES = {
    0: "UNSPECIFIED",
    1: "SCHEDULED",
    2: "STARTED",
    3: "CANCEL_REQUESTED",
}


async def get_workflow_stack_trace(input: GetWorkflowStackTraceInput) -> StackTraceOutput:
    client = await get_temporal_client(input.namespace)

    try:
        handle = client.get_workflow_handle(input.workflow_id, run_id=input.run_id)

        description = await handle.describe()

        stack_trace = ""
        try:
            stack_trace = await handle.query("__stack_trace")
        except Exception:
            stack_trace = "Stack trace unavailable (workflow may not be running or query handler not registered)"

        start_time = description.start_time
        start_time_str = start_time.isoformat() if start_time else datetime.now(timezone.utc).isoformat()

        now = datetime.now(timezone.utc)
        duration_ms = int((now - start_time).total_seconds() * 1000) if start_time else 0

        pending_activities = _extract_pending_activities(description)

        return StackTraceOutput(
            workflow_id=input.workflow_id,
            status=description.status.name if hasattr(description.status, "name") else str(description.status),
            running_since=start_time_str,
            duration_so_far_ms=duration_ms,
            stack_trace=str(stack_trace),
            pending_activities=pending_activities,
        )
    except Exception as e:
        message = str(e)
        if "not found" in message.lower() or "NotFound" in message:
            raise RuntimeError(
                f"No workflow found with ID '{input.workflow_id}'. "
                "The workflow may have been archived or the ID may be incorrect."
            ) from e
        raise RuntimeError(f"Failed to get workflow stack trace: {message}") from e


def _extract_pending_activities(description: Any) -> list[dict[str, Any]]:
    """Extract pending activities from a workflow description."""
    raw = getattr(description, "raw", None) or getattr(description, "raw_description", None)
    if not raw:
        return []

    pending = getattr(raw, "pending_activities", None)
    if not pending:
        return []

    result = []
    for pa in pending:
        activity_type_obj = getattr(pa, "activity_type", None)
        activity_type = getattr(activity_type_obj, "name", "Unknown") if activity_type_obj else "Unknown"

        state = getattr(pa, "state", None)
        if isinstance(state, int):
            state_str = PENDING_ACTIVITY_STATES.get(state, f"UNKNOWN({state})")
        elif state is not None:
            state_str = str(state)
        else:
            state_str = "unknown"

        scheduled_time = getattr(pa, "scheduled_time", None)
        scheduled_time_str = ""
        if scheduled_time:
            if hasattr(scheduled_time, "isoformat"):
                scheduled_time_str = scheduled_time.isoformat()
            elif hasattr(scheduled_time, "seconds"):
                scheduled_time_str = datetime.fromtimestamp(
                    int(scheduled_time.seconds or 0), tz=timezone.utc
                ).isoformat()

        attempt = getattr(pa, "attempt", 1)
        last_failure_obj = getattr(pa, "last_failure", None)
        last_failure = getattr(last_failure_obj, "message", None) if last_failure_obj else None

        result.append({
            "activity_type": activity_type,
            "state": state_str,
            "scheduled_time": scheduled_time_str,
            "attempt": int(attempt) if attempt else 1,
            "last_failure": last_failure,
        })

    return result

File: src/tools/get-workflow-stack-trace.ts
import { getTemporalClient } from "../temporal-client.js";
import type { GetWorkflowStackTraceInput, StackTraceOutput } from "../types.js";

export async function getWorkflowStackTrace(
  input: GetWorkflowStackTraceInput
): Promise<StackTraceOutput> {
  const client = await getTemporalClient(input.namespace);

  try {
    const handle = client.workflow.getHandle(
      input.workflow_id,
      input.run_id
    );

    // Get the workflow description for status and pending activities
    const description = await handle.describe();

    // Query for stack trace (built-in query type)
    let stackTrace = "";
    try {
      stackTrace = await handle.query<string>("__stack_trace");
    } catch {
      stackTrace = "Stack trace unavailable (workflow may not be running or query handler not registered)";
    }

    const startTime = description.startTime
      ? description.startTime.toISOString()
      : new Date().toISOString();

    const now = Date.now();
    const durationMs = description.startTime
      ? now - description.startTime.getTime()
      : 0;

    // Extract pending activities from the raw description
    const pendingActivities = extractPendingActivities(description);

    return {
      workflow_id: input.workflow_id,
      status: description.status.name,
      running_since: startTime,
      duration_so_far_ms: durationMs,
      stack_trace: stackTrace,
      pending_activities: pendingActivities,
    };
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);

    if (message.includes("not found") || message.includes("NotFound")) {
      throw new Error(
        `No workflow found with ID '${input.workflow_id}'. The workflow may have been archived or the ID may be incorrect.`
      );
    }

    throw new Error(`Failed to get workflow stack trace: ${message}`);
  }
}

function extractPendingActivities(
  description: Awaited<ReturnType<import("@temporalio/client").WorkflowHandle["describe"]>>
): StackTraceOutput["pending_activities"] {
  // The WorkflowExecutionDescription from the Temporal client may include
  // raw pendingActivities from the gRPC response
  const raw = description.raw;
  if (!raw?.pendingActivities) return [];

  return raw.pendingActivities.map((pa) => {
    const activityType = pa.activityType?.name || "Unknown";
    const state = pa.state != null
      ? pendingActivityStateToString(pa.state)
      : "unknown";
    const scheduledTime = pa.scheduledTime
      ? new Date(
          Number(pa.scheduledTime.seconds || 0) * 1000
        ).toISOString()
      : "";
    const attempt = pa.attempt || 1;
    const lastFailure = pa.lastFailure?.message || null;

    return {
      activity_type: activityType,
      state,
      scheduled_time: scheduledTime,
      attempt: typeof attempt === "number" ? attempt : Number(attempt),
      last_failure: lastFailure || null,
    };
  });
}

function pendingActivityStateToString(state: number | string): string {
  if (typeof state === "string") return state;

  const stateMap: Record<number, string> = {
    0: "UNSPECIFIED",
    1: "SCHEDULED",
    2: "STARTED",
    3: "CANCEL_REQUESTED",
  };

  return stateMap[state] || `UNKNOWN(${state})`;
}
